### DISCUSSION TITLE
Custom Board with ATtiny826, Programming Questions

### DISCUSSION THREAD

**Discussion by GeorgeIoak** [SpenceKonde/megaTinyCore#580] [Q&A]
I have a new design I'm completing and wanted to be sure I have the correct connections for programming easily. I've gone through many of the pages but I think my head's spinning a bit! Here's the relevant connections I currently have:

![image](https://user-images.githubusercontent.com/1467145/147367548-56c4720d-200c-42c1-9c9a-6d4e48e01459.png)

- The ATtinyx26 parts have the options RESET pin on PB4 but I'm not clear on what to wire this too to take advantage of it
- I have the 2x3 programming header and I know I've assigned pins to the NC pins but that's because I have limited board space. I'm assuming that this is OK and would still work with standard programmers or my own custom adapter?
- If I want to use a serial adapter I'm a little confused on what would give me the most universal connections. If I understand correctly I want a Schottky Diode between Tx and Rx (cathode towards Tx). I assume if I added this to my board I could later use the UART for standard serial interface to another board?
- I have FTDI as well as CP2102 serial adapters. The FTDI doesn't have a series resistor on Tx. Should I add a series resistor on my board or would that make it unusable with a serial adapter that does have a series resistor? I'm hoping to make my design "brain dumb" so you can just grab whatever is on the bench and make it work.
- Since I will be programming new boards do I need to use UPDI to first program in the bootloader? I'm thinking down the road for programming multiple boards and want it to be as quick and easy as possible. 

Thanks in advance for any guidance and Happy Holidays!

**Comment by hmeijdam** [comment]
If i were to design a custom board I would just break out RX/TX as they are, ignore the reset or bootloader (I see no benefit from a bootloader) and break out the UPDI pin with a GND pin as 2-pin programming socket. Or as a 3-pin programming socket with GND/VCC/UPDI 

I made my own dedicated UPDI programmer from a cheapass serial adapter, where I soldered the diode between RX and TX on the adapter. My programmer is powered by the USB port and I power the board separately for programming. No more mistakes feeding 5V to a 3.3V board by accident.

**Reply by GeorgeIoak** [reply]
I was kind of leaning towards not worrying about this board and just making the "custom programmer" like you said. I had room for the 2x3 header so I figured that could be used with standard programmers (like the Atmel ICE) without having to change cables.

Thanks for taking a look.

**Reply by GeorgeIoak** [reply]
This command doesn't seem to have a problem from the command line but it is only reading, `pymcuprog -t uart -u COM5 -b 230400 -d attiny1626 read -m flash -b 4096`

**Reply by GeorgeIoak** [reply]
From the command line I've done `pymcuprog -t uart -u COM5 -b 230400 -d attiny1626 -f readTempVcc.ino.t1626.20c0.mB1.pfF.v254.hex write --verify` 5 times and every time it works without any errors or warnings. As soon as I try it from Arduino I get the `pymcuprog.pymcuprog_errors.PymcuprogError: Error with st` error on the first try

**Reply by apws** [reply]
how you use pymcuprog from arduino? this is board in UPDI mode? I currenlty dont have it under hands... For me flashing UPDI from arduino works great, I had issues with avrdude flashing against optiboot, but there is known bug already (I will probably test also tweak it in commandline, thanks for tip...

**Reply by GeorgeIoak** [reply]
pymcuprog is called when you load the MTC and choose SerialUPDI as your programmer

**Reply by apws** [reply]
ya, but I see now that there is some modified python script, have you latest MTC? this works for me...

C:\Users\petra\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1/python3 -u C:\Users\petra\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.5.4/tools/prog.py -t uart -u COM10 -b 230400 -d attiny1626 --fuses 2:0x02 6:0x04 8:0x00 -fC:\Users\petra\AppData\Local\Temp\arduino_build_496543/megaTinyCore_Blink.ino.hex -a write 
SerialUPDI
UPDI programming for Arduino using a serial adapter
Based on pymcuprog, with significant modifications
By Quentin Bolsee and Spence Konde
Version 1.2.0 - June 2021
Using serial port COM10 at 230400 baud.
Target: attiny1626
Set fuses: ['2:0x02', '6:0x04', '8:0x00']
Action: write
File: C:\Users\petra\AppData\Local\Temp\arduino_build_496543/megaTinyCore_Blink.ino.hex
Pinging device...
Ping response: 1E9429
Setting fuse 0x2=0x2
Writing literal values...
Verifying literal values...
Action took 0.03s
Setting fuse 0x6=0x4
Writing literal values...
Verifying literal values...
Action took 0.03s
Setting fuse 0x8=0x0
Writing literal values...
Verifying literal values...
Action took 0.03s
Finished writing fuses.
Chip/Bulk erase,
Memory type eeprom is conditionally erased (depending upon EESAVE fuse setting)
Memory type flash is always erased
Memory type lockbits is always erased
...
Erased.
Action took 0.01s
Writing from hex file...
Writing flash...
[==================================================] 12/12
Action took 0.15s
Verifying...
[==================================================] 2/2
Verify successful. Data in flash matches data in specified hex-file
Action took 0.05s

**Reply by apws** [reply]
and thank you for tip, for optiboot in latest MTC, there is unparsed params section {serial....} in flashing command, where replacing it by "-PCOM10 -b230400 " in cmdline works ...

**Reply by GeorgeIoak** [reply]
Yes, I have version 2.5.4 MTC installed and it also shows the modified pymcuprog is version 1.2.0. I'm not sure of the modifications that Spence and Quentin made to the program but when I installed `pymcuprog version 3.10.2.133` with pip it ran perfectly with the exact same hardware connections. So for some reason the modifications made programming worse for me but for everyone else it seems the pymcuprog in MTC is better, ...strange.

**Comment by mechatroniks-git** [comment]
[Datasheet ](https://ww1.microchip.com/downloads/en/DeviceDoc/ATtiny424-426-427-824-826-827-DataSheet-DS40002311A.pdf)page 19 recommends 1-10nF bypass in addition to the 100. Optionally, 1uF.  I usually do the 100nF and 1uf. 

The UPDI pin (as you call it reset) should not have pullup or cap on it, at least not without testing. I would expect it to cause UPDI issues.

**Reply by GeorgeIoak** [reply]
take a look at this, https://ww1.microchip.com/downloads/en/DeviceDoc/ETN36_MPLAB%20Snap%20AVR%20Interface%20Modification.pdf with recommendation to add a pull up although that's for the MPLAB SNAP
